frameworklauncher/src/main/java/com/microsoft/frameworklauncher/common/utils/DnsUtils.java

```diff
@@ -17,21 +17,87 @@
 
 package com.microsoft.frameworklauncher.common.utils;
 
+import com.microsoft.frameworklauncher.common.log.DefaultLogger;
+
+import java.net.DatagramSocket;
 import java.net.Inet4Address;
 import java.net.InetAddress;
 import java.net.UnknownHostException;
 
+/**
+ * DnsUtils makes Best Effort to resolve external IP and Host.
+ * An external IP address is the address that other machines is able to connect to it.
+ * Not all IP addresses are reachable outside the local machine or LAN, such as,
+ * the virtual network IP addresses, the private network IP addresses, and so on.
+ */
 public class DnsUtils {
-  // An external address is the address that other machines is able to connect to it.
-  public static String resolveExternalIPv4Address(String hostName) throws UnknownHostException {
-    return Inet4Address.getByName(hostName).getHostAddress();
-  }
+  private static final DefaultLogger LOGGER = new DefaultLogger(DnsUtils.class);
+  private static final String LOCAL_HOST;
+  private static final String LOCAL_IP;
+
+  static {
+    LOGGER.logInfo("Resolving local address");
 
-  public static String resolveLocalHostName() {
+    InetAddress localAddress = null;
+
+    // First use DatagramSocket to resolve, since it will automatically choose
+    // the preferred outbound IP address which should be external.
     try {
-      return InetAddress.getLocalHost().getHostName();
+      DatagramSocket socket = new DatagramSocket();
+
+      // Initialize the local address which will be used to connect with the remote address.
+      // The remote address can be unreachable dummy address since there is no real
+      // connection established here.
+      socket.connect(Inet4Address.getByName("1.2.3.4"), 1234);
+      InetAddress socketLocalAddress = socket.getLocalAddress();
+
+      if (!(socketLocalAddress instanceof Inet4Address) ||
+          socketLocalAddress.isMulticastAddress() ||
+          socketLocalAddress.isLoopbackAddress() ||
+          socketLocalAddress.isAnyLocalAddress()) {
+        throw new Exception(String.format(
+            "SocketLocalAddress [%s] is not usable", socketLocalAddress));
+      }
+
+      localAddress = socketLocalAddress;
     } catch (Exception e) {
-      return null;
+      LOGGER.logWarning(e, "Failed to resolve local address by DatagramSocket");
+    }
+
+    if (localAddress == null) {
+      try {
+        localAddress = Inet4Address.getLocalHost();
```

Comment:
should we consider a case where local IP address may change after the DnsUtil initialized? Your later implementation get the IP on the fly, so doesn't have this issue.

Reply:
good catch. 
It would be better to restart service if its external IP changed. 
Anyway from DnsUtils view, it should not have this assumption. 

---
In reply to: [190150781](https://github.com/Microsoft/pai/pull/601#discussion_r190150781) [](ancestors = 190150781)

Reply:
But from caller sider, even I do not cache here, we still do not have mechanism to detect the ip changes and then updated it accordingly. 
So, let's keep it another work item if future required, ok?

---
In reply to: [190167929](https://github.com/Microsoft/pai/pull/601#discussion_r190167929) [](ancestors = 190167929,190150781)

Reply:
yes. I suggest we just state our assumption here (by adding some comments). no need for further improvement.

---
In reply to: [190170025](https://github.com/Microsoft/pai/pull/601#discussion_r190170025) [](ancestors = 190170025,190167929,190150781)